File: basler/apps/cpp/basler_pipeline.cpp
/*
 * basler_pipeline — Basler camera grab demo
 *
 * Default pipeline (--frames N):
 *   BaslerCameraThread → DumpFrameFilter
 *
 * Grabs N frames from the camera, prints frame info, then exits.
 * Use PYLON_CAMEMU=1 to run without physical hardware.
 *
 * Usage:
 *   export PYLON_CAMEMU=1
 *   ./basler_pipeline [options]
 *
 * Options:
 *   --serial, -s <serial>   Camera serial number (default: first available)
 *   --width,  -w <px>       Capture width  (default: camera default)
 *   --height, -H <px>       Capture height (default: camera default)
 *   --fps,    -f <fps>      Frame rate (default: 30)
 *   --mono,   -m            Use Mono8 mode (default: Color / YUV422)
 *   --frames, -n <count>    Number of frames to grab (default: 30)
 *   --help,   -h
 */

#include <iostream>
#include <atomic>
#include <csignal>
#include <chrono>
#include <thread>
#include <getopt.h>

#include "limef/framefilter/dump.h"
#include "limef/basler/basler_camera_thread.h"

static std::atomic<bool> g_running{true};

static void signal_handler(int sig) {
    if (sig == SIGINT || sig == SIGTERM) {
        std::cout << "\nShutting down..." << std::endl;
        g_running = false;
    }
}

static void print_usage(const char* prog) {
    std::cout
        << "Usage: " << prog << " [options]\n"
        << "\n"
        << "Options:\n"
        << "  --serial, -s <serial>   Camera serial (default: first available)\n"
        << "  --width,  -w <px>       Capture width  (default: camera default)\n"
        << "  --height, -H <px>       Capture height (default: camera default)\n"
        << "  --fps,    -f <fps>      Frame rate (default: 30)\n"
        << "  --mono,   -m            Use Mono8 mode (default: Color)\n"
        << "  --frames, -n <count>    Frames to grab then exit (default: 30, 0=run until Ctrl-C)\n"
        << "  --help,   -h            Show this help\n"
        << "\n"
        << "Test without hardware:\n"
        << "  PYLON_CAMEMU=1 " << prog << "\n"
        << std::endl;
}

int main(int argc, char** argv) {
    std::string serial;
    int  width  = 0;
    int  height = 0;
    double fps  = 30.0;
    bool mono   = false;
    int  frames = 30;

    static struct option long_options[] = {
        {"serial",  required_argument, nullptr, 's'},
        {"width",   required_argument, nullptr, 'w'},
        {"height",  required_argument, nullptr, 'H'},
        {"fps",     required_argument, nullptr, 'f'},
        {"mono",    no_argument,       nullptr, 'm'},
        {"frames",  required_argument, nullptr, 'n'},
        {"help",    no_argument,       nullptr, 'h'},
        {nullptr, 0, nullptr, 0}
    };

    int opt;
    while ((opt = getopt_long(argc, argv, "s:w:H:f:mn:h", long_options, nullptr)) != -1) {
        switch (opt) {
            case 's': serial  = optarg;            break;
            case 'w': width   = std::stoi(optarg); break;
            case 'H': height  = std::stoi(optarg); break;
            case 'f': fps     = std::stod(optarg); break;
            case 'm': mono    = true;              break;
            case 'n': frames  = std::stoi(optarg); break;
            case 'h': print_usage(argv[0]); return 0;
            default:  print_usage(argv[0]); return 1;
        }
    }

    std::signal(SIGINT,  signal_handler);
    std::signal(SIGTERM, signal_handler);

    // ── Build context ─────────────────────────────────────────────────────────

    Limef::basler::BaslerCameraContext ctx;
    ctx.serial        = serial;
    ctx.width         = width;
    ctx.height        = height;
    ctx.fps           = fps;
    ctx.mode          = mono ? Limef::basler::BaslerCameraContext::Mode::Mono
                             : Limef::basler::BaslerCameraContext::Mode::Color;
    ctx.output_format = mono ? AV_PIX_FMT_GRAY8 : AV_PIX_FMT_NV12;

    // ── Build pipeline ────────────────────────────────────────────────────────

    Limef::ff::DumpFrameFilter dump("dump");
    Limef::basler::BaslerCameraThread cam("basler-cam", ctx);

    cam.getOutput().cc(dump);

    // ── Run ───────────────────────────────────────────────────────────────────

    std::cout << "Starting Basler camera ("
              << (mono ? "Mono" : "Color") << " mode)";
    if (!serial.empty()) std::cout << ", serial=" << serial;
    if (width)  std::cout << ", w=" << width;
    if (height) std::cout << ", h=" << height;
    std::cout << ", fps=" << fps;
    if (frames > 0) std::cout << ", grabbing " << frames << " frames";
    std::cout << "\nPress Ctrl-C to stop early.\n" << std::endl;

    cam.start();

    if (frames > 0) {
        // Wait for N frames then stop (simple wall-clock estimate)
        double wait_s = static_cast<double>(frames) / fps + 2.0;
        auto deadline = std::chrono::steady_clock::now()
                      + std::chrono::milliseconds(static_cast<int>(wait_s * 1000));
        while (g_running && std::chrono::steady_clock::now() < deadline) {
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }
    } else {
        while (g_running) {
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }
    }

    cam.stop();
    std::cout << "Done." << std::endl;
    return 0;
}

File: basler/apps/python/basler_pipeline.py
"""
basler_pipeline.py — Basler camera grab demo

Pipeline:
    BaslerCameraThread → DumpFrameFilter

Grabs frames from the camera and prints frame info to the terminal.
Use --frames N to stop after N frames, or Ctrl-C to stop at any time.

Usage:
    python basler_pipeline.py [options]

Options:
    --list     List connected cameras and exit (requires pypylon)
    --serial   Camera serial number (default: first available)
    --width    Capture width  (default: camera default)
    --height   Capture height (default: camera default)
    --fps      Frame rate (default: 30)
    --mono     Use Mono8 mode (default: Color / YUV422)
    --frames   Number of frames to grab then exit (default: 30, 0 = run until Ctrl-C)

Test without hardware:
    PYLON_CAMEMU=1 python basler_pipeline.py
"""

import argparse
import signal
import sys
import time

import limef
import limef_basler


def list_cameras():
    try:
        from pypylon import pylon
    except ImportError:
        print("pypylon is not installed — cannot enumerate cameras.")
        print("Install it with:  pip install pypylon")
        sys.exit(1)

    tl = pylon.TlFactory.GetInstance()
    devices = tl.EnumerateDevices()
    if not devices:
        print("No Basler cameras found.")
        sys.exit(0)

    print(f"{len(devices)} camera(s) found:\n")
    for i, d in enumerate(devices):
        iface = d.GetDeviceClass()          # "BaslerUsb" / "BaslerGigE" / ...
        model  = d.GetModelName()
        serial = d.GetSerialNumber()
        friendly = d.GetFriendlyName()
        line = f"  [{i}]  {model:<28}  serial={serial:<14}  {iface}"
        try:
            ip = d.GetIpAddress()
            if ip and ip != "N/A":
                line += f"  ip={ip}"
        except Exception:
            pass
        print(line)
        if friendly and friendly != model:
            print(f"        friendly name: {friendly}")
    print()
    print("Pass --serial <serial> to select a specific camera.")


def main():
    parser = argparse.ArgumentParser(
        description="Basler camera → DumpFrameFilter demo")
    parser.add_argument("--list",   action="store_true",
                        help="List connected cameras and exit (requires pypylon)")
    parser.add_argument("--serial", default="",
                        help="Camera serial number (default: first available)")
    parser.add_argument("--width",  type=int, default=0,
                        help="Capture width  (default: camera default)")
    parser.add_argument("--height", type=int, default=0,
                        help="Capture height (default: camera default)")
    parser.add_argument("--fps",    type=float, default=30.0,
                        help="Frame rate (default: 30)")
    parser.add_argument("--mono",   action="store_true",
                        help="Use Mono8 mode (default: Color)")
    parser.add_argument("--secs", type=float, default=0,
                        help="Run this many seconds (0 = run until Ctrl-C)")
    parser.add_argument("--feature-file", default="", metavar="PATH",
                        help="Pylon .pfs feature file to load (focus, aperture, exposure, …)")
    parser.add_argument("--no-auto-exposure", action="store_true",
                        help="Disable ExposureAuto (use camera/feature-file setting)")
    args = parser.parse_args()

    if args.list:
        list_cameras()
        sys.exit(0)

    # ── Build pipeline ─────────────────────────────────────────────────────────

    ctx               = limef_basler.BaslerCameraContext()
    ctx.serial        = args.serial
    ctx.width         = args.width
    ctx.height        = args.height
    ctx.fps           = args.fps
    ctx.mode          = limef_basler.Mode.Mono if args.mono else limef_basler.Mode.Color
    ctx.output_format = limef.AV_PIX_FMT_GRAY8 if args.mono else limef.AV_PIX_FMT_NV12
    ctx.feature_file  = args.feature_file
    ctx.exposure_auto = not args.no_auto_exposure

    cam  = limef_basler.BaslerCameraThread("basler-cam", ctx)
    dump = limef.DumpFrameFilter("dump")
    cam.cc(dump)

    # ── Run ────────────────────────────────────────────────────────────────────

    mode_str = "Mono" if args.mono else "Color"
    print(f"Starting Basler camera ({mode_str} mode)"
          + (f", serial={args.serial}" if args.serial else "")
          + (f", {args.width}x{args.height}" if args.width and args.height else "")
          + f", {args.fps} fps"
          + (f", grabbing {args.secs} secs" if args.secs > 0 else "")
    )
    print("Press Ctrl-C to stop early.\n")

    running = True
    def _stop(sig, frame):
        nonlocal running
        running = False
    signal.signal(signal.SIGINT,  _stop)
    signal.signal(signal.SIGTERM, _stop)

    cam.start()

    if args.secs > 0:
        deadline = time.monotonic() + args.secs + 0.1
        while running and time.monotonic() < deadline:
            time.sleep(0.1)
    else:
        while running:
            time.sleep(0.1)

    cam.stop()
    print("Done.")


if __name__ == "__main__":
    main()
